File: card_replacement_demo/card_replacement_demo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
    Homography image replacement demo.
"""

# built-in packages
import argparse
import logging
import os
import string
import sys
import time
import timeit
from functools import partial
# 3rd party packages
import csv
import cv2
import numpy
# custom packages
import util_csv
import util_cv





__author__ = "Eduard Feicho"
__copyright__ = "Copyright 2015"



logger = logging.getLogger(__name__)





class Frame():
    """
        Models a video frame
    """
    def __init__(self, index = 0, name = 'N/A'):
        self.index = index
        self.name  = name




class DemoApp():
    """
        Homography image replacement demo.

        This demo expects as input a folder with an image sequence.
        A reference image and a replacement image of the same size need to be provided (default: reference.jpg, replacement.jpg).
        This is a feature-based image matching demo that works for a planar object.
        The object (reference card) in the video is replaced by the replacement image.
        Think augmented reality.
    """



    def __init__(self):

        # Setup an empty args object, which contains the same variables as the command line argument parser would produce.
        # This way we can either run the python script from command line or alternatively instantiate the class as a module
        # and adjust settings using the args object
        self.args                  = lambda: None
        self.args.logfile          = None
        self.args.verbose          = 0
        self.args.debug            = False

        self.args.input            = './sequence'
        self.args.input_format     = 'seq_%08d.jpg'
        self.args.output           = None
        # self.args.output           = './output'

        self.args.reference        = 'reference.jpg'
        self.args.replacement      = 'replacement.jpg'

        self.args.matcher          = 'flann'


    
    def run(self):
        """
            Main entry point of the app
        """

        ## some initial setup
        
        # parse the input
        self.args.input = self.expandPath(self.args.input)
        self.args.output = self.expandPath(self.args.output)

        # list all frames from directory
        self.names = self.listSequence(path=self.args.input)
        self.paths = self.expandPaths(base=self.args.input, files=self.names)

        # load reference images
        self.im_reference = util_cv.load_image(self.args.reference)
        self.im_replacemt = util_cv.load_image(self.args.replacement)
        self.im_reference_gray = cv2.cvtColor(self.im_reference, cv2.COLOR_BGR2GRAY)
        self.im_replacemt_gray = cv2.cvtColor(self.im_replacemt, cv2.COLOR_BGR2GRAY)
        
        # initialise modules
        self.detector = self.initFeatureDetector()
        self.matcher  = self.initFeatureMatcher()
        
        # compute features for reference image
        self.keypoints_ref, self.descriptors_ref = self.computeFeatures(image=self.im_reference_gray)

        ## main loop (sequential algorithm)
        self.runLoop(self.names, self.paths)



    def runLoop(self, files, paths):
        """
            Process input frame-by-frame
        """

        logger.info(' * Analysing image sequence')

        if len(paths) == 0:
            logger.warn(' Empty sequence, nothing to be done')
            return
        
        # sequence of frames
        frames = []
        for (i, path) in enumerate(paths):

            logger.info('')
            logger.info(' * frame {} : {}'.format(i, path) )

            # set up frame object
            frame = Frame(index=i, name=files[i])
            frames.append(frame)

            # load frame image
            frame.image = util_cv.load_image(path)
            frame.image_gray = cv2.cvtColor(frame.image, cv2.COLOR_BGR2GRAY)

            # detect and compute features
            frame.keypoints, frame.descriptors = self.computeFeatures(image=frame.image_gray)

            # match feature descriptors
            frame.matches = self.matchFeatures(frame.descriptors)

            # debug features
            if self.args.debug:
                debugImage = util_cv.drawMatches(self.im_reference, self.keypoints_ref, frame.image, frame.keypoints, frame.matches)
                util_cv.display_image(debugImage, title='DEBUG')

            ## compute homography from the reference object to the scene
            frame.homography = self.computeHomography(self.keypoints_ref, frame.keypoints, frame.matches)

            # debug homography 
            if self.args.debug:
                debugImage = frame.image.copy()
                util_cv.draw_homography(self.im_reference, frame.homography.astype('float32'), debugImage)
                debugImage = util_cv.drawMatches(self.im_reference, self.keypoints_ref, debugImage, frame.keypoints, frame.matches)
                util_cv.display_image(debugImage, title='DEBUG')


            ## check if homography is bad
            bad_homo = False
            # If the det of the upper 2x2 matrix of the homography is smaller than 0, it is not orientation-preserving.
            # src: http://answers.opencv.org/question/2588/check-if-homography-is-good/
            # This filters out some bad frames, although it's not a perfect criteria (tested with the bad results using BruteForceMatcher)
            det2x2 = cv2.determinant(frame.homography[0:2,0:2])
            if det2x2 < 0:
                bad_homo = True

            # This criteria can be found in the image stitching pipeline of opencv...
            det3x3 = abs(cv2.determinant(frame.homography))
            if det3x3 < 0.00001:#sys.float_info.epsilon:
                bad_homo = True

            if bad_homo:
                
                # homography is useless, just show the original picture
                frame.im_replacemt = frame.image.copy()

            else:

                # warp replacement image and merge with the frame
                h,w = frame.image.shape[:2]
                im_warped = cv2.warpPerspective(self.im_replacemt, frame.homography, dsize=(w,h))
                im_replacemt = numpy.ma.array(frame.image, mask=im_warped > 0, fill_value=0).filled()
                im_replacemt = cv2.bitwise_or(im_replacemt, im_warped)
                frame.im_replacemt = im_replacemt


            # debug result
            if self.args.debug:
                color = (255,0,0)
                if det2x2 < 0 or det3x3 < 0.00001:
                    color = (0,0,255)
                cv2.putText(frame.im_replacemt, 'Frame {}: {}'.format(i, os.path.basename(path)), (0,25), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0,0,0))
                cv2.putText(frame.im_replacemt, 'det2x2 = {}'.format(det2x2), (0,50), cv2.FONT_HERSHEY_SIMPLEX, 0.5, color)
                cv2.putText(frame.im_replacemt, 'det3x3 = {}'.format(det3x3), (0,75), cv2.FONT_HERSHEY_SIMPLEX, 0.5, color)
                util_cv.display_images([frame.image, frame.im_replacemt], title='Processed Frame')

            # write image
            if self.args.output:
                output = path.replace(self.args.input, self.args.output)
                logger.info( ' * Write {}'.format(output) )
                cv2.imwrite( output, frame.im_replacemt )

            # clear memory
            frames.pop(0)



    def listSequence(self, path):
        """
            Given a path, assume it's a folder and return a sorted list of the folder contents
        """

        logger.info(' * Read input sequence from directory: {}'.format(path))

        if not os.path.isdir(path):
            logger.error('path is not a directory: {}'.format(path))
            return []

        files = os.listdir(path)
        # TODO filter only valid entries using self.args.input_format
        files = filter(lambda file: ".jpg" in file, files)
        files = sorted(files)

        logger.info(' * Found {} files'.format(len(files)))

        return files



    def expandPath(self, path):
        if path is None:
            return None
        return os.path.abspath(os.path.expanduser(path))



    def expandPaths(self, base, files):
        """
            expand user home folder and convert to absolute paths
        """
        files = map(lambda p: self.expandPath( os.path.join(base, p) ), files)
        return files



    def initFeatureDetector(self):
        return cv2.SURF(hessianThreshold=400, nOctaves=4, nOctaveLayers=2, extended=True, upright=False)



    def initFeatureMatcher(self):
        if self.args.matcher == 'bruteforce':
            return cv2.BFMatcher(normType=cv2.NORM_L2)
        else:
            
            # bug: flann enums are missing
            # workaround based on src: http://stackoverflow.com/questions/8301962/opencv-pythons-api-flannbasedmatcher

            FLANN_INDEX_KDTREE = 1
            indexParams = dict(algorithm = FLANN_INDEX_KDTREE, trees = 4)
            return cv2.FlannBasedMatcher(indexParams=indexParams, searchParams={})



    def computeFeatures(self, image):
        """
            Detect interesting keypoints and compute a set of features
        """
        keypoints, descriptors = self.detector.detectAndCompute(image, mask=None)
        return keypoints, descriptors



    def matchFeatures(self, features_query):
        """
            Match feature vectors against reference features
        """
        if self.args.matcher == 'flann':
            # bug: flann enums are missing
            matches = self.matcher.knnMatch(self.descriptors_ref, features_query, k=2)
            # store all the good matches as per Lowe's ratio test.
            matches = self.applyRatioTest(matches)
        else:
            matches = self.matcher.match(self.descriptors_ref, features_query)

        return matches





    def applyRatioTest(self, matches, ratio=0.75):
        """
            Ratio test from D.Lowe's SIFT paper
        """

        # Apply ratio test
        good = []
        for m,n in matches:
            if m.distance < ratio*n.distance:
                good.append(m)
        return good



    def computeHomography(self, keypoints_src, keypoints_dst, good_matches):
        """
            Compute homography given keypoints and matches
        """
        # Localize the object

        src_pts = numpy.float32([ keypoints_src[m.queryIdx].pt for m in good_matches ]).reshape(-1,1,2)
        dst_pts = numpy.float32([ keypoints_dst[m.trainIdx].pt for m in good_matches ]).reshape(-1,1,2)
        homography, mask = cv2.findHomography( src_pts, dst_pts, cv2.cv.CV_RANSAC, 5.0)

        logger.debug('H={}'.format(homography))

        return homography





def main():

    # initialize the app (contains default values for some command line arguments)
    app = DemoApp()

    # setup command line argument parser
    parser = argparse.ArgumentParser(description="Homography image replacement demo.")
    parser.add_argument("input",                metavar="INPUT_FOLDER",                                                              help="Path to input sequence folder")
    parser.add_argument("-o", "--output",       dest="output",          default=app.args.output,                                     help="Path to output sequence folder")
    parser.add_argument("--reference",          dest="reference",       default=app.args.reference,                                  help="Path to reference image")
    parser.add_argument("--replacement",        dest="replacement",     default=app.args.replacement,                                help="Path to replacement image")
    parser.add_argument("--format",             dest="input_format",    default=app.args.input_format,                               help="Format of input sequence")
    parser.add_argument("--matcher",            dest="matcher",         default=app.args.matcher,                                    help="Matcher (flann|bruteforce|...)")
    
    # various arguments related to logging/debugging
    parser.add_argument("-l", "--logfile",      dest="logfile",         default=None,                                                help="Path to logfile")
    parser.add_argument("-v", "--verbose",      dest="verbose",         default=1,      action="count",                              help="verbosity level (use -vv to increase more)")

    parser.add_argument("-q", "--quiet",        dest="quiet",           default=False,  action="store_true",                         help="don't print status messages to stdout")
    #parser.add_argument("--test",               dest="test",            default=False,  action="store_true",                         help="run unit tests (TODO)")
    parser.add_argument("--debug",              dest="debug",           default=False,  action="store_true",                         help="enable debugging mode")
    parser.add_argument("--evaluate",           dest="evaluate",        default=False,  action="store_true",                         help="enable evaluation mode")

    # parse command line arguments
    args = parser.parse_args()
    if args.quiet: args.verbose = 0
    if args.debug: args.verbose = 2

    # setup logging from command line arguments
    setupLogging(args, logger)

    # use command line arguments to specify the configuration of the app
    app.args = args

    # the main entry point of the app
    p = partial(app.run)
    time_seconds = timeit.timeit( p, number=1)
    logger.info(' * Algorithm took {:.2f} seconds for {:5d} frames ({:.2f} FPS)'.format(time_seconds, len(app.names), float(len(app.names))/time_seconds))



def setupLogging(args, logger, basicFormat='%(levelname)s:\t%(message)s', fileFormat='%(asctime)s\t%(levelname)s:\t%(message)s', loglevel=None):
    """
        a basic logging setup based on some assumptions of the given command line arguments
    """

    # if logging level is not explcitly specified, chose based on verbosity level specified on command line
    if loglevel is None:
        loglevel = logging.WARNING                            # default to log level WARNING
        if args.verbose >= 1: loglevel = logging.INFO         # specify for example -v or --verbose
        if args.verbose >= 2: loglevel = logging.DEBUG        # specify for example -vv

    logging.basicConfig(format=basicFormat, level=loglevel)

    # write to file, if -l/--logfile specified on command line
    if args.logfile is not None:
        handler = logging.FileHandler(args.logfile)
        formatter = logging.Formatter(fileFormat)
        handler.setLevel(loglevel)
        handler.setFormatter(formatter)
        logger.addHandler(handler)



if __name__ == "__main__":
    main()

File: card_replacement_demo/card_replacement_demo.cpp
/*
 * Homography image replacement demo.
 */



// std libs
#include <iostream>
#include <string>
using namespace std;


#include <cfloat>       // defines FLT_EPSILON

// custom libs
#include <opencv2/core/core.hpp>
#include <opencv2/calib3d/calib3d.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/features2d/features2d.hpp>
#include <opencv2/nonfree/features2d.hpp>
// using namespace cv;  // predilection









/********************************************************************************
 * @TODO: A quick and dirty logger
 ********************************************************************************/
extern int log_level;
int log_level = 0;
#define LOG_INFO(msg)           if (log_level > 0) { std::cout << " INFO :\t" << msg << std::endl; }
#define LOG_DEBUG(msg)          if (log_level > 1) { std::cout << " DEBUG:\t" << msg << std::endl; }
#define LOG_ERROR(msg)          std::cerr << " ERROR:\t" << msg << std::endl;
#define LOG_WARN(msg)           std::cerr << " WARN :\t" << msg << std::endl;






// C-like string formatting
// src: http://stackoverflow.com/questions/2342162/stdstring-formatting-like-sprintf
template <typename... Ts>
std::string fmt(const std::string &fmt, Ts... vs)
{
    char b;
    unsigned required = std::snprintf(&b, 0, fmt.c_str(), vs...) + 1;
        // See comments: the +1 is necessary, while the first parameter
        //               can also be set to nullptr

    char bytes[required];
    std::snprintf(bytes, required, fmt.c_str(), vs...);

    return std::string(bytes);
}






/********************************************************************************
 * opencv utility functions
 ********************************************************************************/


namespace util_cv
{

/**
 * Load image located at given path
 */
cv::Mat load_image(const string& path, int options = CV_LOAD_IMAGE_COLOR)
{
    LOG_DEBUG("Loading " << path);
    cv::Mat mat = cv::imread(path, options);
    if (mat.empty())
    {
        LOG_ERROR("Image empty for file: " << path);
    }
    return mat;
}



/**
  Scales an image then displays it using OpenCV's inbuilt methods.

    :param img: input image
    :param title: window title
    :param wait: waitKey?
    :param windowScale: scale image
 */
void display_image(const cv::Mat& img, const string& title = "Debug", const bool wait = true, const float windowScale = 1.0, const bool close = false, const bool createWindow = true)
{
    cv::Mat tmp = img;
    if (windowScale != 1.0)
    {
        int height = int(round(img.rows * windowScale));
        int width  = int(round(img.cols * windowScale));
        cv::resize(img, tmp, cv::Size(width, height));
    }
    
    if (createWindow)
    {
        cv::namedWindow(title);
    }
    cv::imshow(title, tmp);
    if (wait)
    {
        cv::waitKey();
    }
    if (close)
    {
        cv::destroyWindow(title);
    }
}



/**
  Stacks a set of images horizontally before displaying

    :param img: input image
    :param title: window title
    :param wait: waitKey?
    :param windowScale: scale image
 */
void display_images(const vector<cv::Mat>& images, const string title = "Debug", const bool wait = true, const float windowScale = 1.0, const bool close = false, const bool createWindow = true)
{
    // @TODO support non-equally-sized images by padding zeros

    if (images.size() == 0)
    {
        return;
    }

    // stack images horizontally
    auto it_end = end(images);
    auto it = begin(images);
    cv::Mat tmp = (*it++);
    while (it != it_end)
    {
        cv::hconcat(tmp, (*it++), tmp);
    }
    display_image(tmp, title, wait, windowScale, close, createWindow);
}




void displayKeypoints(const cv::Mat& full_img, const vector<cv::KeyPoint>& keypoints)
{
    cv::namedWindow("keypoints", 1);
    cv::Mat img_keypoints;
    cv::drawKeypoints(full_img, keypoints, img_keypoints);
    cv::imshow("keypoints", img_keypoints);
    cv::waitKey(0);
}


void displayMatches(const cv::Mat &img_1, const std::vector<cv::KeyPoint> &keypoints_1, const cv::Mat &img_2, const std::vector<cv::KeyPoint> &keypoints_2, const std::vector<cv::DMatch> &matches, cv::Mat &img_matches)
{
    //-- Draw matches
    cv::drawMatches( img_1, keypoints_1, img_2, keypoints_2, matches, img_matches );
    //-- Show detected matches
    cv::namedWindow( "Matches", CV_WINDOW_NORMAL );
    cv::imshow("Matches", img_matches );
    cv::waitKey(0);
}




void draw_homography(const cv::Mat& img1, const cv::Mat& H, const cv::Mat &img2)
{
    int h = img1.rows;
    int w = img1.cols;
    
    std::vector<cv::Point2f> obj_corners(4);
    obj_corners[0] = cvPoint(   0, 0);
    obj_corners[1] = cvPoint( w-1, 0 );
    obj_corners[2] = cvPoint( w-1, h-1 );
    obj_corners[3] = cvPoint(   0, h-1 );
    std::vector<cv::Point2f> scene_corners(4);
    std::vector<cv::Point2i> scene_corners_int;
    cv::perspectiveTransform(obj_corners, scene_corners, H);

    for (auto pt : scene_corners)
    {
        scene_corners_int.push_back( cv::Point2i(pt.x, pt.y) );
    }

    copy(begin(scene_corners), end(scene_corners), inserter(scene_corners_int, end(scene_corners_int)));
    cv::polylines(img2, scene_corners_int, true, cv::Scalar(0, 255, 0), 3, CV_AA);
}



};






/********************************************************************************
 * Datastructures
 ********************************************************************************/




/**
 * Data structure for a video frame
 */
class Frame
{
public:
    Frame(int index_, string name_)
    :
    index(index_),
    name(name_)
    {
    }

    // unique image ID, such as frame number
    int     index;
    // a possible user-defined name
    string  name;

    // the image data in BGR
    cv::Mat image;
};


/**
 * Datastructure for a video frame and (temporary) derived data
 */
class ProcessedFrame : public Frame
{
public:
    ProcessedFrame(int index, string name)
    :
    Frame(index,name)
    {
    }
    
    // the frame image as grayscale
    cv::Mat image_gray;
    // the final output image where the detected planar reference object is replaced with the replacement image
    cv::Mat image_replacement;

    // vector of 2d interest point locations
    vector<cv::KeyPoint> keypoints;
    // vector of feature descriptors for each interest point
    cv::Mat descriptors;
    // vector of feature matches
    vector<struct cv::DMatch> matches;
    // homography for the detected reference object
    cv::Mat homography;
};








/********************************************************************************
 * Main Demo Code
 ********************************************************************************/




/**
    Homography image replacement demo.

    This demo expects as input a folder with an image sequence.
    A reference image and a replacement image of the same size need to be provided (default: reference.jpg, replacement.jpg).
    This is a feature-based image matching demo that works for a planar object.
    The object (reference card) in the video is replaced by the replacement image.
    Think augmented reality.
 */
class DemoApp
{

public:
    /********************************************************************************
     * Datastructures
     ********************************************************************************/

    enum class MatcherType
    {
        BRUTEFORCE,
        FLANN
    };
    enum class DetectorType
    {
        SIFT,
        SURF,
        ORB,
        FAST,
        CUSTOM
    };
    enum class FeaturesType
    {
        SIFT,
        SURF,
        ORB,
        CUSTOM
    };

    typedef struct t_options
    {
        // verbosity count
        int verbose = 0;
        // toggle debuggin mode
        bool debug = false;

        // output directory for results
        string output = "./output";

        // reference image (planar object that is to be detected)
        string reference = "reference.jpg";
        // replacement image for the reference image in the scene
        string replacement = "replacement.jpg";

        // support different feature detectors, descriptors and matchers
        DetectorType detector_type = DetectorType::SURF;
        FeaturesType features_type = FeaturesType::SURF;
        MatcherType matcher_type   = MatcherType::FLANN;
    } Options;



private:
    /********************************************************************************
     * Class members
     ********************************************************************************/

    // holds all options of this demo
    Options _options;

    // list all frames from stdin
    vector<string> _paths;

    
    cv::Mat _im_reference;
    cv::Mat _im_replacemt;
    cv::Mat _im_reference_gray;
    cv::Mat _im_replacemt_gray;

    cv::Ptr<cv::FeatureDetector>        _detector;
    cv::Ptr<cv::DescriptorExtractor>    _extractor;
    cv::Ptr<cv::DescriptorMatcher>      _matcher;

    vector<cv::KeyPoint> _keypoints_ref;
    cv::Mat _descriptors_ref;


public:

    /********************************************************************************
     * Logic
     ********************************************************************************/
    DemoApp()
    {
    }


    DemoApp(Options& options)
    :
    _options(options)
    {
        // empty function
    }

    

public:

    /********************************************************************************
     * Public API
     ********************************************************************************/

    // Main entry point of the app
    void run()
    {
        /* some initial setup */

        // list all frames from stdin
        _paths = listSequence();

        // load reference images
        _im_reference = util_cv::load_image(_options.reference);
        _im_replacemt = util_cv::load_image(_options.replacement);
        cv::cvtColor(_im_reference, _im_reference_gray, cv::COLOR_BGR2GRAY);
        cv::cvtColor(_im_replacemt, _im_replacemt_gray, cv::COLOR_BGR2GRAY);

        if (_options.debug)
        {
            util_cv::display_images(vector<cv::Mat>( {_im_reference, _im_replacemt} ), "Reference/Replacement");
        }

        
        // initialise modules
        _detector  = initFeatureDetector();
        _extractor = initFeatureExtractor();
        _matcher   = initFeatureMatcher();

        
        /* pre-compute features for referenc images */
        _keypoints_ref = detectKeypoints(_im_reference_gray);
        _descriptors_ref = computeFeatures(_im_reference_gray, _keypoints_ref);
        

        /* main loop (sequential algorithm) */
        runLoop(_paths);
    }


    // Process input frame-by-frame
    void runLoop(vector<string> &paths)
    {
        LOG_INFO(" * Analysing image sequence");
        
        const int N = paths.size();

        if (N == 0)
        {
            LOG_WARN(" Empty sequence, nothing to be done");
            return;
        }
        
        int frame_count = -1;
        for (const auto& path : paths)
        {
            ++frame_count;


            LOG_INFO("");
            LOG_INFO(" * frame " << frame_count << " : " << path);

            // set up frame object
            ProcessedFrame frame(frame_count, path);

            // load frame image
            frame.image = util_cv::load_image(path);
            cv::cvtColor(frame.image, frame.image_gray, cv::COLOR_BGR2GRAY);

            // detect and compute features
            frame.keypoints   = detectKeypoints(frame.image_gray);
            frame.descriptors = computeFeatures(frame.image_gray, frame.keypoints);
            
            // match feature descriptors
            frame.matches = matchFeatures(frame.descriptors);

            // debug features
            if (_options.debug)
            {
                cv::Mat debugImage;
                cv::drawMatches(_im_reference, _keypoints_ref, frame.image, frame.keypoints, frame.matches, debugImage);
                util_cv::display_image(debugImage, "DEBUG");
            }

            // compute homography from the reference object to the scene
            frame.homography = computeHomography(_keypoints_ref, frame.keypoints, frame.matches);


            // debug homography
            if (_options.debug)
            {
                cv::Mat debugImage = frame.image.clone();
                frame.homography.convertTo(frame.homography, CV_32FC1);
                util_cv::draw_homography(_im_reference, frame.homography, debugImage);
                cv::drawMatches(_im_reference, _keypoints_ref, debugImage.clone(), frame.keypoints, frame.matches, debugImage);
                util_cv::display_image(debugImage, "DEBUG");
            }

            // check if homography is bad
            bool bad_homography = false;
            // If the det of the upper 2x2 matrix of the homography is smaller than 0, it is not orientation-preserving.
            // src: http://answers.opencv.org/question/2588/check-if-homography-is-good/
            // This filters out some bad frames, although it's not a perfect criteria (tested with the bad results using BruteForceMatcher)
            double det2x2 = cv::determinant( frame.homography(cv::Rect(0,0,2,2)).clone() );
            if (det2x2 < 0)
            {
                bad_homography = true;
            }

            // This criteria can be found in the image stitching pipeline of opencv...
            double det3x3 = abs( cv::determinant(frame.homography) );
            if (det3x3 < 0.00001) // alternative: FLT_EPSILON
            {
                bad_homography = true;
            }

            if (bad_homography)
            {
                // homography is useless, just show the original picture
                frame.image_replacement = frame.image.clone();
            }
            else
            {
                // warp replacement image and merge with the frame
                int h = frame.image.rows;
                int w = frame.image.cols;
                cv::Mat im_warped;
                cv::warpPerspective(_im_replacemt, im_warped, frame.homography, cv::Size(w, h));

                frame.image_replacement = frame.image.clone();
                im_warped.copyTo(frame.image_replacement, im_warped > 0);
            }


            // debug result
            if (_options.debug)
            {
                cv::Scalar color = cv::Scalar(255, 0, 0);
                if (det2x2 < 0 || det3x3 < 0.00001) // alt: FLT_EPSILON
                {
                    color = cv::Scalar(0, 0, 255);
                }

                cv::putText(frame.image_replacement, fmt("Frame %d: %s", frame_count, path.c_str()), cv::Point2i(0,25), cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0,0,0));
                cv::putText(frame.image_replacement, fmt("det2x2 = %f", det2x2), cv::Point2i(0,50), cv::FONT_HERSHEY_SIMPLEX, 0.5, color);
                cv::putText(frame.image_replacement, fmt("det3x3 = %f", det3x3), cv::Point2i(0,75), cv::FONT_HERSHEY_SIMPLEX, 0.5, color);
                util_cv::display_images(vector<cv::Mat>( {frame.image, frame.image_replacement} ), "Processed Frame");
            }

            // write image
            if (_options.output.length() != 0)
            {
                string output = fmt("%s/%s", _options.output.c_str(), path.substr( path.find_last_of('/')+1 ).c_str());
                LOG_INFO( " * Write " << output );
                cv::imwrite( output, frame.image_replacement );
            }
        }
    }



    /*
     * Parse STDIN and assume each line contains a file path
     */
    vector<string> listSequence()
    {
        LOG_INFO(" * Read input sequence from STDIN ");

        vector<string> files;
        files.reserve(50);
        string line;
        while (std::getline(std::cin, line))
        {
            files.push_back(line);
        }

        LOG_INFO(" * Found " << files.size() << " files");

        return files;
    }


    /*
     * Initialise keypoint detector
     */
    cv::Ptr<cv::FeatureDetector> initFeatureDetector()
    {
        cv::Ptr<cv::FeatureDetector> featureDetector;
        
        switch (_options.detector_type)
        {
        case DetectorType::ORB:
            featureDetector = cv::FeatureDetector::create("PyramidORB");
            break;
        case DetectorType::FAST:
            featureDetector = cv::FeatureDetector::create("PyramidFAST");
            break;
        case DetectorType::CUSTOM:
            featureDetector = new cv::PyramidAdaptedFeatureDetector(
                new cv::DynamicAdaptedFeatureDetector(
                    // new cv::FastAdjuster(20, true), 30, 50, 20),
                   new cv::FastAdjuster(20, true), 5000, 10000, 10),
                    8
                );
            break;
        case DetectorType::SIFT:
            featureDetector = cv::FeatureDetector::create("SIFT");
            break;
        case DetectorType::SURF: // default
        default:
            featureDetector = cv::FeatureDetector::create("PyramidSURF");
            break;
        }

        return featureDetector;
    }


    /*
     * Initialise descriptor extractor
     */
    cv::Ptr<cv::DescriptorExtractor> initFeatureExtractor()
    {
        cv::Ptr<cv::DescriptorExtractor> extractor;

        switch (_options.features_type)
        {
        case FeaturesType::ORB:
            extractor = cv::DescriptorExtractor::create("ORB");
            break;
        case FeaturesType::SIFT:
            extractor = cv::DescriptorExtractor::create("SIFT");
            break;

        case FeaturesType::CUSTOM: // TODO
        case FeaturesType::SURF: // default
        default:
            extractor = cv::DescriptorExtractor::create("SURF");
            break;
        }

        return extractor;
    }


    /*
     * Initialise the feature matcher
     */
    cv::Ptr<cv::DescriptorMatcher> initFeatureMatcher()
    {
        cv::Ptr<cv::DescriptorMatcher> matcher;
        if (_options.matcher_type == MatcherType::BRUTEFORCE)
        {
            const int normType = cv::NORM_L2;

            matcher = new cv::BFMatcher(normType);
        }
        else
        {
            const cv::Ptr<cv::flann::IndexParams>& indexParams = new cv::flann::KDTreeIndexParams();
            const cv::Ptr<cv::flann::SearchParams>& searchParams = new cv::flann::SearchParams();

            matcher = new cv::FlannBasedMatcher(indexParams, searchParams);
        }
        return matcher;
    }



    /*
     * Detect interesting keypoints
     */
    vector<cv::KeyPoint> detectKeypoints(const cv::Mat& image)
    {
        vector<cv::KeyPoint> keypoints;
        (*_detector).detect(image, keypoints);
        return keypoints;
    }



    /*
     * Describe interesting keypoints
     */
    cv::Mat computeFeatures(const cv::Mat& image, vector<cv::KeyPoint>& keypoints)
    {
        cv::Mat descriptors;
        (*_extractor).compute(image, keypoints, descriptors);
        return descriptors;
    }



    /*
     * Match feature vectors against reference features
     */
    vector<struct cv::DMatch> matchFeatures(const cv::Mat& features_query)
    {
        vector< vector<struct cv::DMatch> > matches;
        vector<struct cv::DMatch> good_matches;
        if (_options.matcher_type == MatcherType::FLANN)
        {
            // bug: flann enums are missing
            (*_matcher).knnMatch(_descriptors_ref, features_query, matches, 2);
            // store all the good matches as per Lowe's ratio test.
            good_matches = applyRatioTest(matches);
        }
        else
        {
            (*_matcher).match(_descriptors_ref, features_query, good_matches);
        }

        return good_matches;
    }




    /*
     * Ratio test from D.Lowe's SIFT paper
     */
    vector<struct cv::DMatch> applyRatioTest(vector< vector<struct cv::DMatch> >& matches, const float& ratio = 0.75)
    {
        /*
            An alternate method of determining high-quality feature matches is the ratio test
            proposed by David Lowe in his paper on SIFT (page 20 for an explanation).
            This test rejects poor matches by computing the ratio between the best and
            second-best match. If the ratio is below some threshold, the match is discarded as being low-quality.
            
            src: http://stackoverflow.com/questions/17967950/improve-matching-of-feature-points-with-opencv
        */

        vector<struct cv::DMatch> good;
        good.reserve(matches.size());
        int i = 0;
        for (auto v_match : matches)
        {
            if (v_match[0].distance < ratio * v_match[1].distance)
            {
                good.push_back(v_match[0]);
            }
        }
        
        return good;
    }



    /*
     * Compute homography given keypoints and matches
     */
    cv::Mat computeHomography(const vector<cv::KeyPoint>& keypoints_src, const vector<cv::KeyPoint>& keypoints_dst, const vector<struct cv::DMatch>& good_matches)
    {
        const int N = good_matches.size();

        cv::vector<cv::Point2f> src_pts;
        cv::vector<cv::Point2f> dst_pts;
        src_pts.reserve(N);
        dst_pts.reserve(N);
        for (auto match : good_matches)
        {
            src_pts.push_back( keypoints_src[match.queryIdx].pt );
            dst_pts.push_back( keypoints_dst[match.trainIdx].pt );
        }

        cv::Mat mask;
        cv::Mat homography = cv::findHomography( src_pts, dst_pts, mask, CV_RANSAC, 5.0);

#if 0
        // TODO: This two-stage optimisation is not really better (seq_00000071.jpg).. why?..

        cv::vector<cv::Point2f> src_inliers;
        cv::vector<cv::Point2f> dst_inliers;
        src_inliers.reserve(N);
        dst_inliers.reserve(N);
        
        for (int i=0; i<mask.rows; ++i)
        {
            auto inlier = mask.at<double>(i,0);
            if (inlier > 0)
            {
                src_inliers.push_back( src_pts[i] );
                dst_inliers.push_back( dst_pts[i] );
            }
        }

        homography = cv::findHomography( src_inliers, dst_inliers, CV_LMEDS);
#endif

        LOG_DEBUG("H=" << homography);

        return homography;
    }
};




int main(int argc, char** argv)
{
    // initialize the app (contains default values for some command line arguments)
    log_level = 2;
    DemoApp::Options options;
    options.debug = true;
    DemoApp app(options);
    app.run();
}
